lib/src/Uncertainty/Algorithm/Sensitivity/openturns/HSICEstimatorImplementation.hxx

```diff
@@ -0,0 +1,179 @@
+#include "openturns/Matrix.hxx"
+#include "openturns/Graph.hxx"
+#include "openturns/PersistentCollection.hxx"
+#include "openturns/ResourceMap.hxx"
+#include "openturns/CovarianceModel.hxx"
+#include "openturns/SymbolicFunction.hxx"
+#include "openturns/Function.hxx"
+#include "openturns/HSICStat.hxx"
+#include "openturns/RandomGenerator.hxx"
+BEGIN_NAMESPACE_OPENTURNS
+
+/**
+ * @class HSICEstimatorImplementation
+ *
+ */
+class OT_API HSICEstimatorImplementation
+  : public PersistentObject
+{
+  CLASSNAME
+public:
+
+  typedef Collection <CovarianceModel>  CovarianceModelCollection;
+
+public:
+
+  /** Default constructor */
+  HSICEstimatorImplementation();
+
+  /** Constructor with a weight function */
+  HSICEstimatorImplementation(const CovarianceModelCollection & covarianceList, const Sample & X, const Sample & Y, const HSICStat & estimatorType);
+
+  /** Constructor with a unitary weight function */
+  HSICEstimatorImplementation(const CovarianceModelCollection & covarianceList, const Sample & X, const Sample & Y, const HSICStat & estimatorType, const Function & weightFunction );
+
+  /* Here is the interface that all derived class must implement */
+public:
+  /** Virtual constructor */
+  HSICEstimatorImplementation * clone() const override;
+
+protected:
+  /** Compute the weight matrix from the weight function */
+  virtual SquareMatrix computeWeightMatrix(const Sample & Y) const;
+
+  /** Compute a HSIC index (one marginal) by using the underlying estimator (biased or not) */
+  virtual Scalar computeHSICIndex( const Sample & inSample, const Sample & outSample, const CovarianceModel & inCovariance, const CovarianceModel & outCovariance, const SquareMatrix & weightMatrix) const;
+
+  /** Compute HSIC and R2-HSIC indices */
+  virtual void computeIndices() const;
+
+public:
+  /** Set the number of permutation used */
+  void setPermutationSize(const UnsignedInteger B);
+
+  /** Get the number of permutation used */
+  UnsignedInteger getPermutationSize() const;
+
+  /** Get the covariance models */
+  CovarianceModelCollection getCovarianceList() const;
+
+  /** Set the covariance models */
+  void setCovarianceList(const CovarianceModelCollection & coll);
+
+  /** Get the input sample */
+  Sample getInputSample() const;
+
+  /** Set the input sample */
+  void setInputSample(const Sample & inputSample);
+
+  /** Get the output sample */
+  Sample getOutputSample() const;
+
+  /** Set the output sample */
+  void setOutputSample(const Sample & outputSample);
+
+  /** Get the dimension of the indices: the number of marginals */
+  UnsignedInteger getDimension() const;
+
+  /** Get the size of the study sample */
+  UnsignedInteger getSize() const;
+
+  /** Get the underlying estimator: biased or unbiased*/
+  HSICStat getEstimator() const;
+
+protected:
+  /** Reset indices to void */
+  void resetIndices();
+
+protected:
+  /** Compute p-value with permutation */
+  virtual void computePValuesPermutation() const;
+
+public:
+
```

Comment:
usually we put public stuff first, then protected, then private, it makes the class easier to read

Reply:
OK.